### PR TITLE
time enhancements - reset epoch, wrapped millis

### PR DESCRIPTION
**Pull request by jepler**
* At soft reset, always reset the monotonic epoch.  This means that   a "fresh" session will not have the accuracy problems that result   when `time.monotonic()` becomes a large float and the precision   decreases

 * A new function `reset_monotonic_epoch` in supervisor can reset the   epoch to now; Careful use within a program can prevent `time.monotonic()`   from losing "too much" accuracy.

 * A new function `ticks_ms` in supervisor returns the _possibly wrapped_   time since the epoch.  The wrapping number is chosen for the reasons   given in #3410: math on these values can be performed in Python   without the need to create long integer values (and thus avoids   allocations and can work on platforms without longs)

   It could be argued whether `ticks_ms` should have its epoch reset with   the others.  I chose not to do so, because programs should be written   so that they function properly when `ticks_ms` wraps

### CODE REVIEW

File: shared-bindings/supervisor/__init__.c [hunk 1]
```diff
@@ -207,6 +208,26 @@ STATIC mp_obj_t supervisor_set_next_code_file(size_t n_args, const mp_obj_t *pos
...
+//|     """Reset the epoch used by the monotonic clock functions."""
+//|     ...
+//|
+STATIC mp_obj_t supervisor_reset_monotonic_epoch(void) {
+    common_hal_time_reset_monotonic_epoch();
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_0(supervisor_reset_monotonic_epoch_obj, supervisor_reset_monotonic_epoch);
+
+//| def ticks() -> int:
```
Comment (dhalbert) [inline shared-bindings/supervisor/__init__.c:221]:
```suggestion
//| def ticks_ms() -> int:
```
to match the name

**Review by dhalbert — CHANGES_REQUESTED** [review]
This is a simple set of additions and is elegant. Just one doc typo.

File: shared-bindings/supervisor/__init__.c [hunk 2]
```diff
@@ -218,11 +218,11 @@ STATIC mp_obj_t supervisor_reset_monotonic_epoch(void) {
...
 MP_DEFINE_CONST_FUN_OBJ_0(supervisor_reset_monotonic_epoch_obj, supervisor_reset_monotonic_epoch);
 
-//| def ticks() -> int:
-//|     """Return the time in milliseconds since an unknown reference point, wrapping after `(1<<29)`ms"""
+//| def ticks_ms() -> int:
+//|     """Return the time in milliseconds since an unknown reference point, wrapping after 2**29ms"""
 STATIC mp_obj_t supervisor_ticks_ms(void) {
     uint64_t ticks_ms = common_hal_time_monotonic_ms();
-    return mp_obj_new_int(ticks_ms % (1 << 29));
+    return mp_obj_new_int((ticks_ms + 0x1fff7777) % (1 << 29));
```
Comment (dhalbert) [inline shared-bindings/supervisor/__init__.c:225]: I think it would be good to explain this in a comment. It seems like a pretty random number.

**Review by dhalbert — CHANGES_REQUESTED** [review]

**Comment by jepler** [comment]
This filled up a couple of m0 builds (non-adafruit products).  I turned off rotaryio to let them fit.

**Comment by jepler** [comment]
we may want to decide whether resetting "monotonic" or providing a wrapping ticks_ms is a better use of limited flash, rather than trying to do both.

**Pull request closed without merging**